### ISSUE TITLE
Imagenet100 Reproduction Accuracy considerably lower than reported.

### ISSUE THREAD

**Issue by davidrzs** [vturrisi/solo-learn#346]
**Describe the bug**
Might not be a bug but your guidance is valued:

I tried to reproduce the imagenet100 results using Barlow Twins and get accuracies well below the officially reported figures.
![image](https://github.com/vturrisi/solo-learn/assets/18198078/2d1a20a8-a01b-4ded-9038-f8c19e6955b0)
The model trained for 400 epochs with the default configuration on 2 GPUs, but even when just training on 1 GPU the same accuracies can be observed. I suspect that 400 epochs is not enough (as the validation curve keeps increasing), hence I wanted to check if you really trained the models for 400 epochs. 

I suspect I am doing something wrong but I am unsure what my mistake is. Thanks for your guidance. 

For the dataset: We used [these](https://github.com/HobbitLong/CMC/blob/master/imagenet100.txt) classes as referenced [here](https://github.com/vturrisi/solo-learn/issues/116).

**To Reproduce**
`python /ssl_pm/main_pretrain.py --config-path scripts/pretrain/imagenet-100/ --config-name barlow.yaml`

**Screenshots**
![image](https://github.com/vturrisi/solo-learn/assets/18198078/29a1d5e8-40cd-45f5-a7da-86b7ac09bc54)

**Versions**
Latest version (pulled 2 days ago)

-- 
here you have my `barlow.yaml` file for reference:

```yaml
defaults:
  - _self_
  - augmentations: asymmetric.yaml
  - wandb: private.yaml
  - override hydra/hydra_logging: disabled
  - override hydra/job_logging: disabled

# disable hydra outputs
hydra:
  output_subdir: null
  run:
    dir: .

name: "barlow_twins-imagenet100"
method: "barlow_twins"
backbone:
  name: "resnet18"
method_kwargs:
  proj_hidden_dim: 2048
  proj_output_dim: 2048
  scale_loss: 0.1
data:
  dataset: imagenet100
  train_path: "/itet-stor/zdavid/net_scratch/ILSVRC2012_img_train_100" #"/home/david/data/imagenet_100/train"
  val_path: "/itet-stor/zdavid/net_scratch/ILSVRC2012_img_val_100" #"/home/david/data/imagenet_100/val"
  format: "dali"
  num_workers: 4
optimizer:
  name: "lars"
  batch_size: 128
  lr: 0.3
  classifier_lr: 0.1
  weight_decay: 1e-4
  kwargs:
    clip_lr: True
    eta: 0.02
    exclude_bias_n_norm: True
scheduler:
  name: "warmup_cosine"
checkpoint:
  enabled: True
  dir: "trained_models"
  frequency: 1
auto_resume:
  enabled: False

# overwrite PL stuff
max_epochs: 400
devices: [0, 1]
sync_batchnorm: True
accelerator: "gpu"
strategy: "ddp"
precision: 16-mixed
```

**Comment by vturrisi** [comment]
Hi @davidrzs. Thanks for letting us know.

Can you try to run the previous version, with an older pytorch lightning version, etc? This should work as reported in the results table.

The much lower accuracy can be due to some pytorch lightning bug/version change or was introduced by some change that I did in the latest version. The only thing that comes to my mind is that the LARS that I introduced to the repo is messed up.

**Comment by DonkeyShot21** [comment]
The curve you posted is really smooth, usually you should have a bit more variance. It's very possibly an optimization problem as Victor suggested. Can you send the loss as well?

**Comment by vturrisi** [comment]
I said LARS but actually we didn't change it, we just changed the LR Scheduler. I don't think it introduced any bug, but can you also share the LR plot?

**Comment by vturrisi** [comment]
There are two other things that come to my mind. DALI could have changed something. Can you also try without it? Second thing is that training now uses 16-mixed precision instead of the old 16 precision. I noticed that for notebook GPU (1660ti) it simply doesn't work and the outputs of the model are just NaNs. It doesn't give any error or anything like that. Can you try full precision training (this worked for me)?

**Comment by davidrzs** [comment]
Here are some screenshots: 
![image](https://github.com/vturrisi/solo-learn/assets/18198078/264c78ac-ea2d-46b9-a00c-ea17d2fac644)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/09057d8b-e7c3-4584-a952-02ebf8a1a5f6)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/1fcc998d-9b7a-470f-8415-6c0e09c8d279)

I can schedule a 32bit run later tonight or tomorrow with DALI disabled (for the old packages it will take some time as getting these virtual environments to work on the cluster is a real pain due to memory restrictions).

Looking at the learning rate plot, I assume that something is a bit odd there, as we do not have the full progression that we see with other models. 

Let me know if any other screenshots or data would be helpful.

**Comment by vturrisi** [comment]
@davidrzs Thanks for all the screenshots. Is this the full training? If so, there's definitely an issue with the LR scheduler and most likely that's the cause of the low performance. It can be that lightning is now calling the scheduler in a different way, I don't know.

**Comment by davidrzs** [comment]
Yes, this is the full 400 epochs (i.e. the run exited completely).

**Comment by vturrisi** [comment]
The batch size in the yaml file is per GPU, that's fine.

**Comment by vturrisi** [comment]
Me and @DonkeyShot21 were talking and this can be an issue that happened quite a long time ago with Dali + pytorch lightning. This is the issue https://github.com/NVIDIA/DALI/issues/3865. I'll need some time to check again, but if you see what works from the following, it would be great:
- Dali + single GPU
- No Dali + single GPU
- No Dali and multi GPU
Just need to run for like 20 epochs so we go past the warmup of the LR Scheduler. We shouldn't see any epoch skipping nor issues with the LR plot.

**Comment by davidrzs** [comment]
I already tested single GPU and Dali, same behaviour: (here both single GPU and the above multi GPU are plotted):

![image](https://github.com/vturrisi/solo-learn/assets/18198078/ce132096-01c7-40a3-a14a-de112b29372c)

Will try a "non-Dali" run and report back.

**Comment by davidrzs** [comment]
This is from a non-Dali run on two Titan Xp cards:

In the images you can see the previous with Dali run and the new still running non-Dali run:

![image](https://github.com/vturrisi/solo-learn/assets/18198078/3e11d232-d397-4fb5-9e03-1a92b11dd3fe)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/637012a5-58dc-4a1d-aa1f-17d75f5fd7f9)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/14f3c17e-3350-4856-ba06-217d6cd8880c)

Do I interpret the epoch chart correctly, that somehow the Dali run gets twice the number of epochs per global trainer step? (which would be weird)

**Comment by vturrisi** [comment]
If the issue is the same as before, Dali calls some reset method after each epoch, which then triggers an "epoch skip" from lightning. This looks very much to be the case. I'll check what is the proper fix in the next couple of days.

**Comment by davidrzs** [comment]
I let it run overnight (still running, will let it complete) and we are now around epoch 160: 

The following are the graphs: 

![image](https://github.com/vturrisi/solo-learn/assets/18198078/166070d7-1c99-4267-95c2-4012702a9103)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/daaab14b-959c-4130-ac34-1e35a68d2abb)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/0fe8782c-b3fb-4e0f-886f-ade1e6cf1edf)

I feel like the accuracy is still lower than where it should be given that soon half of the training will be done?

**Comment by vturrisi** [comment]
@davidrzs I checked my old runs, and it seems quite similar to the curves that I have.
![Screenshot_20230528-113844.png](https://github.com/vturrisi/solo-learn/assets/34046216/e877bbf2-8178-4a10-a9ea-74232160f032)

**Comment by davidrzs** [comment]
I have just scheduled the same model on the cluster with Dali enabled (will report in due time).

The last run (without Dali) terminated after 380 epochs (we have a 48h limit on the cluster) and the following are the plots:

![image](https://github.com/vturrisi/solo-learn/assets/18198078/92ef4dad-6da7-49a0-8b29-dba655dd6995)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/c1673455-851d-4741-ae49-986063063292)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/1b2db2f0-1aa3-400d-9427-95b8345ff239)

![image](https://github.com/vturrisi/solo-learn/assets/18198078/2c98d225-b377-4968-b3c9-0391686d6eff)

Top1-train seems to be okay, but top1 val seems to be way too low. Do you have any idea why? Will investigate on Wednesday a bit.

**Comment by vturrisi** [comment]
Hummm. It can be the change that I added to comply with lightning>=2.0. I'll check that soon (if that's the case, it's only an evaluation issue).

**Comment by davidrzs** [comment]
Can confirm that it works with Dali on pytorch lightning==1.9.0. (2gpus)

See screenshots:


![image](https://github.com/vturrisi/solo-learn/assets/18198078/e03ba31f-4e03-4cd6-afb5-7e7caf6f2201)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/d12e6f3a-afba-4e04-bc10-1f7f16fbd4b9)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/f9db2bcf-1e97-4e0f-bf2b-aa47e82725c2)


Though, the evaluation issue persists. Do you have any idea what could cause the eval issue? That way I could also start to investigate.

**Comment by vturrisi** [comment]
@davidrzs I'll push an update that will fix DALI for lightning>2.0. About the accuracy issue, lightning changed how you compute epoch-wise metrics. Beforehand, you just returned it in the `validation_step` and you would receive it as parameter in another callback. Now, you need to store the metrics in a list and remember manually clean it. I simply forgot to call `.clear()` on it. The push will also fix this.

**Comment by davidrzs** [comment]
Thank you very much for your work on this! Will rerun it after you merged with main and report tomorrow.

**Comment by vturrisi** [comment]
Merged it. The modification is really simple so hopefully, it's fine.

**Comment by davidrzs** [comment]
Can confirm that it works (learning rate as well as validation accuracies).
Thanks for fixing it!

![image](https://github.com/vturrisi/solo-learn/assets/18198078/6736ef09-ac06-4b80-88dd-934a8ac62407)
![image](https://github.com/vturrisi/solo-learn/assets/18198078/cabc2bc5-5ef4-4f2e-b0c5-ae7a4fe716fc)


Though, you might want to update pytorch lightning in the requirements file. (I manually updated and let it run on 2.0.2

**Comment by vturrisi** [comment]
Super! It's always annoying to update library versions, but I'm glad it's working now :). I'll also update the requirements file, thought I already had.
